### PR TITLE
enhancement(performance): Adjust the strength of `gauge_add` atomics

### LINKED ISSUE #6477: Audit atomics and use weakest ordering possible
As discussed in https://github.com/timberio/vector/pull/6428#discussion_r577777055 we should use `compare_exchange_weak` where possible. It's worth taking a passes through the code to find opportunities for weaker ordering.

### PR DESCRIPTION
**Pull request by blt**
This commit adjusts the strength of the `gauge_add` atomics used by
`OpenGauge`. Previous implementation called `emitter` potentially `n^2 / 2`
times, where `n` is the number of participants. Current implementation calls
`emitter` only once after the participant CPU manages to adjust `OpenGauge`
value, though no guarantee is made about the ordering of participant calls to
`emitter`. That is, this commit guarantees that thread `A` will sequence its call
to `emitter` after modification of the internal `gauge` but makes no guarantee
about sequence of thread `B`'s parallel call to its `emitter`, relative to `A`.

If the point of this code is to sequence calls to `emitter` between threads we
can close this out and I'll make a different PR to clarify in the
documentation.

Closes #6477

### CODE REVIEW

**Review by bruceg — APPROVED** [review]
LGTM

**Review by jszwedko — APPROVED** [review]
Thanks @blt ! This looks good to me. I'm a bit confused why were were calling the `emitter` on every loop even before this change.

**Review by ktff — CHANGES_REQUESTED** [review]
> If the point of this code is to sequence calls to emitter between threads

Yep, that's goal. The main reason is to guarantee

>  but makes no guarantee
about sequence of thread B's parallel call to its emitter, relative to A.

since without it a gauge metric with older value can overwrite a gauge metric with newer value until a third update comes, which could take a while since connection aren't opened/closed all the time, and during that time internal metrics will report old value.

**Comment by blt** [comment]
@ktff appreciate the clarification. Will address.

**Comment by blt** [comment]
@ktff 5dbe9a7 should address your feedback. Please do take a look when you have a chance.

File: src/internal_events/open.rs [hunk 1]
```diff
@@ -60,27 +61,40 @@ impl<E: Fn(usize)> Drop for OpenToken<E> {
...
+        // Try to update gauge to new value and releasing writes to gauge metric
+        // in the process.  Otherwise acquire new writes to gauge metric.
+        //
+        // When `compare_exchange_weak` returns Ok our `new_value` is now the
+        // current value in memory across all CPUs. When the return is Err we
+        // retry with the now current value.
+        match gauge.compare_exchange_weak(value, new_value, Ordering::AcqRel, Ordering::Acquire) {
+            Ok(_) => break,
+            Err(x) => {
+                hint::spin_loop();
```
Comment (ktff) [inline src/internal_events/open.rs:93]: Nice

**Review by ktff — APPROVED** [review]
Nice, on the point, .... well it seams I'll need to write a book to gain this ability.

File: src/internal_events/open.rs (hunk 1 above)
Comment (blt) [reply src/internal_events/open.rs:93]: :bow:

**Pull request merged**